In [1]:
# Price Data

item_dict = {'almonds': 6,
 'antioxidant juice': 9,
 'avocado': 1,
 'black tea': 4,
 'burgers': 4,
 'champagne': 3,
 'chicken': 4,
 'chocolate': 7,
 'chutney': 7,
 'cookies': 7,
 'cooking oil': 7,
 'cottage cheese': 9,
 'eggs': 9,
 'energy bar': 9,
 'energy drink': 3,
 'extra dark chocolate': 8,
 'french fries': 7,
 'fresh tuna': 9,
 'frozen smoothie': 7,
 'frozen vegetables': 2,
 'green grapes': 8,
 'green tea': 8,
 'honey': 9,
 'light cream': 5,
 'low fat yogurt': 4,
 'meatballs': 9,
 'milk': 5,
 'mineral water': 2,
 'oil': 9,
 'olive oil': 8,
 'pet food': 3,
 'protein bar': 1,
 'salad': 7,
 'salmon': 3,
 'shallot': 6,
 'shrimp': 2,
 'soup': 2,
 'spaghetti': 6,
 'spinach': 7,
 'tomato juice': 8,
 'tomatoes': 5,
 'turkey': 8,
 'vegetables mix': 3,
 'whole wheat flour': 5,
 'whole wheat pasta': 5,
 'whole wheat rice': 4,
 'yams': 7}

In [2]:
# transaction data

transaction_data = \
[['shrimp', 'almonds', 'avocado', 'vegetables mix', 'green grapes', 'whole wheat flour', 'yams', 'cottage cheese', 'energy drink', 'tomato juice', 'low fat yogurt', 'green tea', 'honey', 'salad', 'mineral water', 'salmon', 'antioxidant juice', 'frozen smoothie', 'spinach', 'olive oil'],
['burgers', 'meatballs', 'eggs'],
['chutney'],
['turkey', 'avocado'],
['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'],
['low fat yogurt'],
['whole wheat pasta', 'french fries'],
['soup', 'light cream', 'shallot'],
['frozen vegetables', 'spaghetti', 'green tea'],
['french fries'],
['eggs', 'pet food'],
['cookies'],
['turkey', 'burgers', 'mineral water', 'eggs', 'cooking oil'],
['spaghetti', 'champagne', 'cookies'],
['mineral water', 'salmon'],
['mineral water'],
['shrimp', 'chocolate', 'chicken', 'honey', 'oil', 'cooking oil', 'low fat yogurt'],
['turkey', 'eggs'],
['turkey', 'fresh tuna', 'tomatoes', 'spaghetti', 'mineral water', 'black tea', 'salmon', 'eggs', 'chicken', 'extra dark chocolate'],
['meatballs', 'milk', 'honey', 'french fries', 'protein bar']]

In [3]:
cost_dict = item_dict.copy()
cost_dict

{'almonds': 6,
 'antioxidant juice': 9,
 'avocado': 1,
 'black tea': 4,
 'burgers': 4,
 'champagne': 3,
 'chicken': 4,
 'chocolate': 7,
 'chutney': 7,
 'cookies': 7,
 'cooking oil': 7,
 'cottage cheese': 9,
 'eggs': 9,
 'energy bar': 9,
 'energy drink': 3,
 'extra dark chocolate': 8,
 'french fries': 7,
 'fresh tuna': 9,
 'frozen smoothie': 7,
 'frozen vegetables': 2,
 'green grapes': 8,
 'green tea': 8,
 'honey': 9,
 'light cream': 5,
 'low fat yogurt': 4,
 'meatballs': 9,
 'milk': 5,
 'mineral water': 2,
 'oil': 9,
 'olive oil': 8,
 'pet food': 3,
 'protein bar': 1,
 'salad': 7,
 'salmon': 3,
 'shallot': 6,
 'shrimp': 2,
 'soup': 2,
 'spaghetti': 6,
 'spinach': 7,
 'tomato juice': 8,
 'tomatoes': 5,
 'turkey': 8,
 'vegetables mix': 3,
 'whole wheat flour': 5,
 'whole wheat pasta': 5,
 'whole wheat rice': 4,
 'yams': 7}

# Q1. Based on customer demand, items keep changing in the store. Your first task is to create a function that lets the admin add or remove items from the store directory(Cost dictionary).

1. Find how many unique items are present in the data?  

In [4]:
#unique items in a dictionary can be found by its length
print("No of unique items are:", len(item_dict))

No of unique items are: 47


In [5]:
#alternate method-by traversing the dictionary
l1 = []
count = 0
for item in item_dict:
    if item not in l1:
        count += 1
        l1.append(item)

print("No of unique items are:", count)

No of unique items are: 47


2. Find the total number of transactions?

In [6]:
#total number of transaction can be found by its length
print("Total number of transactions is", len(transaction_data))

Total number of transactions is 20


3. Add items, try adding the item “banana” in stores. The cost of this item is $5. Add the item “banana” to the Cost dictionary. 

In [7]:
#adding new item to cost dictionary
cost_dict['banana'] = 5
cost_dict

{'almonds': 6,
 'antioxidant juice': 9,
 'avocado': 1,
 'black tea': 4,
 'burgers': 4,
 'champagne': 3,
 'chicken': 4,
 'chocolate': 7,
 'chutney': 7,
 'cookies': 7,
 'cooking oil': 7,
 'cottage cheese': 9,
 'eggs': 9,
 'energy bar': 9,
 'energy drink': 3,
 'extra dark chocolate': 8,
 'french fries': 7,
 'fresh tuna': 9,
 'frozen smoothie': 7,
 'frozen vegetables': 2,
 'green grapes': 8,
 'green tea': 8,
 'honey': 9,
 'light cream': 5,
 'low fat yogurt': 4,
 'meatballs': 9,
 'milk': 5,
 'mineral water': 2,
 'oil': 9,
 'olive oil': 8,
 'pet food': 3,
 'protein bar': 1,
 'salad': 7,
 'salmon': 3,
 'shallot': 6,
 'shrimp': 2,
 'soup': 2,
 'spaghetti': 6,
 'spinach': 7,
 'tomato juice': 8,
 'tomatoes': 5,
 'turkey': 8,
 'vegetables mix': 3,
 'whole wheat flour': 5,
 'whole wheat pasta': 5,
 'whole wheat rice': 4,
 'yams': 7,
 'banana': 5}

4. Use a function named add_item( ) to automatically add an item to the Cost dictionary given its name and price.

In [8]:
#function to add new items to cost dictionary given its name and price
def add_item(dict_name,name,price):
    dict_name[name] = price
    return dict_name

In [9]:
#driver code for add_item() function with an example
add_item(cost_dict, 'red banana', 5)

{'almonds': 6,
 'antioxidant juice': 9,
 'avocado': 1,
 'black tea': 4,
 'burgers': 4,
 'champagne': 3,
 'chicken': 4,
 'chocolate': 7,
 'chutney': 7,
 'cookies': 7,
 'cooking oil': 7,
 'cottage cheese': 9,
 'eggs': 9,
 'energy bar': 9,
 'energy drink': 3,
 'extra dark chocolate': 8,
 'french fries': 7,
 'fresh tuna': 9,
 'frozen smoothie': 7,
 'frozen vegetables': 2,
 'green grapes': 8,
 'green tea': 8,
 'honey': 9,
 'light cream': 5,
 'low fat yogurt': 4,
 'meatballs': 9,
 'milk': 5,
 'mineral water': 2,
 'oil': 9,
 'olive oil': 8,
 'pet food': 3,
 'protein bar': 1,
 'salad': 7,
 'salmon': 3,
 'shallot': 6,
 'shrimp': 2,
 'soup': 2,
 'spaghetti': 6,
 'spinach': 7,
 'tomato juice': 8,
 'tomatoes': 5,
 'turkey': 8,
 'vegetables mix': 3,
 'whole wheat flour': 5,
 'whole wheat pasta': 5,
 'whole wheat rice': 4,
 'yams': 7,
 'banana': 5,
 'red banana': 5}

5. Remove an item: for example: if the customers at Walmart do not seem to be purchasing “banana,” and the store decides to discontinue this item. The function to should be able to remove the item “banana” from the Cost dictionary

In [10]:
#deleting an item
del cost_dict['banana']
cost_dict

{'almonds': 6,
 'antioxidant juice': 9,
 'avocado': 1,
 'black tea': 4,
 'burgers': 4,
 'champagne': 3,
 'chicken': 4,
 'chocolate': 7,
 'chutney': 7,
 'cookies': 7,
 'cooking oil': 7,
 'cottage cheese': 9,
 'eggs': 9,
 'energy bar': 9,
 'energy drink': 3,
 'extra dark chocolate': 8,
 'french fries': 7,
 'fresh tuna': 9,
 'frozen smoothie': 7,
 'frozen vegetables': 2,
 'green grapes': 8,
 'green tea': 8,
 'honey': 9,
 'light cream': 5,
 'low fat yogurt': 4,
 'meatballs': 9,
 'milk': 5,
 'mineral water': 2,
 'oil': 9,
 'olive oil': 8,
 'pet food': 3,
 'protein bar': 1,
 'salad': 7,
 'salmon': 3,
 'shallot': 6,
 'shrimp': 2,
 'soup': 2,
 'spaghetti': 6,
 'spinach': 7,
 'tomato juice': 8,
 'tomatoes': 5,
 'turkey': 8,
 'vegetables mix': 3,
 'whole wheat flour': 5,
 'whole wheat pasta': 5,
 'whole wheat rice': 4,
 'yams': 7,
 'red banana': 5}

Create a function called remove_item( ) to automatically remove an item from the Cost dictionary given its name.

In [11]:
#function to remove item from cost dictionary
def remove_item(dict_name,prod):
    del dict_name[prod]
    return dict_name

In [12]:
#drivers code for remove_item() function
remove_item(cost_dict, 'red banana')

{'almonds': 6,
 'antioxidant juice': 9,
 'avocado': 1,
 'black tea': 4,
 'burgers': 4,
 'champagne': 3,
 'chicken': 4,
 'chocolate': 7,
 'chutney': 7,
 'cookies': 7,
 'cooking oil': 7,
 'cottage cheese': 9,
 'eggs': 9,
 'energy bar': 9,
 'energy drink': 3,
 'extra dark chocolate': 8,
 'french fries': 7,
 'fresh tuna': 9,
 'frozen smoothie': 7,
 'frozen vegetables': 2,
 'green grapes': 8,
 'green tea': 8,
 'honey': 9,
 'light cream': 5,
 'low fat yogurt': 4,
 'meatballs': 9,
 'milk': 5,
 'mineral water': 2,
 'oil': 9,
 'olive oil': 8,
 'pet food': 3,
 'protein bar': 1,
 'salad': 7,
 'salmon': 3,
 'shallot': 6,
 'shrimp': 2,
 'soup': 2,
 'spaghetti': 6,
 'spinach': 7,
 'tomato juice': 8,
 'tomatoes': 5,
 'turkey': 8,
 'vegetables mix': 3,
 'whole wheat flour': 5,
 'whole wheat pasta': 5,
 'whole wheat rice': 4,
 'yams': 7}

# Q2. Once a customer has completed shopping, they need to be billed for the total amount of all the items. This can be done automatically by swiping each item’s barcode on a checkout system. Your second task is to build the software underlying the checkout system, which finds the total cost of all the items in a customer’s cart, including taxes:

1. Write the code, to find the price of an item from the cost dictionary

    Find the price of salmon from the cost dictionary

In [13]:
#to find cost of user input item from cost dictionary
x = input("Enter the item name : ") #taking item name as input from user
print("Price of {0} is {1}".format(x,cost_dict[x]))

Enter the item name : salmon
Price of salmon is 3


In [14]:
# alternate method
cost_dict['salmon']

3

2. Write a code that helps the administrator find any particular transaction in the sequence

    Try finding the ninth transaction from the data

In [15]:
#to find the items in user input transaction number
x = int(input("Enter the transaction number : ")) #taking transaction number as input from user
print("transaction number {0} is {1}".format(x,transaction_data[x-1]))

Enter the transaction number : 9
transaction number 9 is ['frozen vegetables', 'spaghetti', 'green tea']


In [16]:
#alternate method
transaction_data[8]

['frozen vegetables', 'spaghetti', 'green tea']

3. How many items are present in the ninth transaction?

In [17]:
#to find the number of items in user input transaction number
x = int(input("Enter the transaction number : ")) #taking transaction number as input from user
print("Total number of items in transaction number {0} is {1} ".format(x,len(transaction_data[x-1])))

Enter the transaction number : 9
Total number of items in transaction number 9 is 3 


In [18]:
#alternate method
len(transaction_data[8])

3

4. Find the total cost of the ninth transaction from the array of transactions.

In [19]:
#function to find total cost of user input transaction number
d = {} #creating empty dictionary
for i in range(len(transaction_data)):#looping through transaction data
    total = 0
    for j in transaction_data[i]:
        total+= item_dict[j]
    d[i+1] = total

x = int(input("Enter the transaction number : ")) #taking transaction number input from user
print("Total cost of transaction number {0} is {1}".format(x,d[x]))

Enter the transaction number : 9
Total cost of transaction number 9 is 16


5. As per government guidelines, a mandatory 5% tax is applied to the total cost of a transaction. Find the final price of the second transaction after applicable taxes have been implemented.

In [33]:
#function to find final cost of user input transaction number
d = {} #creating empty dictionary
for i in range(len(transaction_data)):   #looping through the transaction data
    total = 0
    for j in transaction_data[i]:
        total+= item_dict[j]
    d[i+1] = total


x = int(input("Enter the transaction number : ")) #take transaction number as input from user
t = d[x]
tax = 5
f = t + ((tax*t)/100) #formula to add tax to the cost
print("Final cost of transaction number {0} after tax is {1}".format(x,f))

Enter the transaction number : 2
Final cost of transaction number 2 after tax is 23.1


6. Create a function named find_total( ) to find the final price of any transaction after applicable taxes have been implemented.

In [21]:
#function to find the final cost of user input transaction number
def find_total(n):
    d = {}
    for i in range(len(transaction_data)):
        total = 0
        for j in transaction_data[i]:
            total+= item_dict[j]
        d[i+1] = total 
        
    t = d[n]
    f = t + ((5*t)/100)
    print("Final cost of transaction number {0} is {1}".format(n,f))

#driver code for above function
find_total(2)

Final cost of transaction number 2 is 23.1


# Q3. There is always a flow of items in and out of stores, which need to be tracked so that the store can place an order in case an item is running out of stock. Your third task is to build an inventory management system that will count the total number of items purchased from the transaction data. This inventory management system will go through each of the transactions one by one and increment the value for an item if it is found in a transaction.

1. What is the data structure that can be used to count the frequency of items purchased (from the transaction data)?

ANS: Data structure such as dictionary can be used

2. Create a frequency counter to count the number of items:
 

In [22]:
frequency_dict = {}  # creating the empty dictionary

for transaction in transaction_data: # looping over each transaction
    for item in transaction: # looping over each item present in the transaction
        if item in frequency_dict.keys():
            frequency_dict[item] +=1 # increment the frequency by 1 if the item is found in the dictionary
        else:
            frequency_dict[item] = 1 # create a new key with the item name if the item is found for the first time
    
    
#printing the dictionary with frequency counter
print(frequency_dict)

{'shrimp': 2, 'almonds': 1, 'avocado': 2, 'vegetables mix': 1, 'green grapes': 1, 'whole wheat flour': 1, 'yams': 1, 'cottage cheese': 1, 'energy drink': 1, 'tomato juice': 1, 'low fat yogurt': 3, 'green tea': 3, 'honey': 3, 'salad': 1, 'mineral water': 6, 'salmon': 3, 'antioxidant juice': 1, 'frozen smoothie': 1, 'spinach': 1, 'olive oil': 1, 'burgers': 2, 'meatballs': 2, 'eggs': 5, 'chutney': 1, 'turkey': 4, 'milk': 2, 'energy bar': 1, 'whole wheat rice': 1, 'whole wheat pasta': 1, 'french fries': 3, 'soup': 1, 'light cream': 1, 'shallot': 1, 'frozen vegetables': 1, 'spaghetti': 3, 'pet food': 1, 'cookies': 2, 'cooking oil': 2, 'champagne': 1, 'chocolate': 1, 'chicken': 2, 'oil': 1, 'fresh tuna': 1, 'tomatoes': 1, 'black tea': 1, 'extra dark chocolate': 1, 'protein bar': 1}


3. A new transaction has just been made. This transaction includes “honey,” “pet food,” and “shrimp.” Update the frequency counter to reflect this new transaction.

In [23]:
transaction_list = transaction_data.copy() #creating a copy of original list
new_list = ['honey', 'pet food','shrimp']
transaction_list.append(new_list) # appending the new transaction to the list
print(transaction_list)

[['shrimp', 'almonds', 'avocado', 'vegetables mix', 'green grapes', 'whole wheat flour', 'yams', 'cottage cheese', 'energy drink', 'tomato juice', 'low fat yogurt', 'green tea', 'honey', 'salad', 'mineral water', 'salmon', 'antioxidant juice', 'frozen smoothie', 'spinach', 'olive oil'], ['burgers', 'meatballs', 'eggs'], ['chutney'], ['turkey', 'avocado'], ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'], ['low fat yogurt'], ['whole wheat pasta', 'french fries'], ['soup', 'light cream', 'shallot'], ['frozen vegetables', 'spaghetti', 'green tea'], ['french fries'], ['eggs', 'pet food'], ['cookies'], ['turkey', 'burgers', 'mineral water', 'eggs', 'cooking oil'], ['spaghetti', 'champagne', 'cookies'], ['mineral water', 'salmon'], ['mineral water'], ['shrimp', 'chocolate', 'chicken', 'honey', 'oil', 'cooking oil', 'low fat yogurt'], ['turkey', 'eggs'], ['turkey', 'fresh tuna', 'tomatoes', 'spaghetti', 'mineral water', 'black tea', 'salmon', 'eggs', 'chicken', 'extra 

In [24]:
#function to update the frequency counter
def update_counter():
    frequency_dict = {}  # creating the empty dictionary
    for transaction in transaction_list: # looping over each transaction
        for item in transaction: # looping over each item present in the transaction
            if item in frequency_dict:
                frequency_dict[item] +=1 # increment the frequency by 1 if the item is found in the dictionary
            else:
                frequency_dict[item] = 1# create a new key with the item name if the item is found for the first time
    
    print(frequency_dict)

update_counter() #calling the function

{'shrimp': 3, 'almonds': 1, 'avocado': 2, 'vegetables mix': 1, 'green grapes': 1, 'whole wheat flour': 1, 'yams': 1, 'cottage cheese': 1, 'energy drink': 1, 'tomato juice': 1, 'low fat yogurt': 3, 'green tea': 3, 'honey': 4, 'salad': 1, 'mineral water': 6, 'salmon': 3, 'antioxidant juice': 1, 'frozen smoothie': 1, 'spinach': 1, 'olive oil': 1, 'burgers': 2, 'meatballs': 2, 'eggs': 5, 'chutney': 1, 'turkey': 4, 'milk': 2, 'energy bar': 1, 'whole wheat rice': 1, 'whole wheat pasta': 1, 'french fries': 3, 'soup': 1, 'light cream': 1, 'shallot': 1, 'frozen vegetables': 1, 'spaghetti': 3, 'pet food': 2, 'cookies': 2, 'cooking oil': 2, 'champagne': 1, 'chocolate': 1, 'chicken': 2, 'oil': 1, 'fresh tuna': 1, 'tomatoes': 1, 'black tea': 1, 'extra dark chocolate': 1, 'protein bar': 1}


4. Create a function named update_counter( ) to automatically update the frequency counter in case any new transaction is made.

In [25]:
#function to update counter
def update_counter():
    frequency_dict = {}  # creating the empty dictionary
    for transaction in transaction_list: # looping over each transaction
        for item in transaction: # looping over each item present in the transaction
            if item in frequency_dict:
                frequency_dict[item] +=1 # increment the frequency by 1 if the item is found in the dictionary
            else:
                frequency_dict[item] = 1# create a new key with the item name if the item is found for the first time
    
    print(frequency_dict)
    

#calling the function
update_counter()

{'shrimp': 3, 'almonds': 1, 'avocado': 2, 'vegetables mix': 1, 'green grapes': 1, 'whole wheat flour': 1, 'yams': 1, 'cottage cheese': 1, 'energy drink': 1, 'tomato juice': 1, 'low fat yogurt': 3, 'green tea': 3, 'honey': 4, 'salad': 1, 'mineral water': 6, 'salmon': 3, 'antioxidant juice': 1, 'frozen smoothie': 1, 'spinach': 1, 'olive oil': 1, 'burgers': 2, 'meatballs': 2, 'eggs': 5, 'chutney': 1, 'turkey': 4, 'milk': 2, 'energy bar': 1, 'whole wheat rice': 1, 'whole wheat pasta': 1, 'french fries': 3, 'soup': 1, 'light cream': 1, 'shallot': 1, 'frozen vegetables': 1, 'spaghetti': 3, 'pet food': 2, 'cookies': 2, 'cooking oil': 2, 'champagne': 1, 'chocolate': 1, 'chicken': 2, 'oil': 1, 'fresh tuna': 1, 'tomatoes': 1, 'black tea': 1, 'extra dark chocolate': 1, 'protein bar': 1}


In [26]:
#alternate method to update counter automatically when adding new transaction
def add_transaction(transaction):
    for item in transaction:
            if item in frequency_dict:
                frequency_dict[item] +=1 # increment the frequency by 1 if the item is found in the dictionary
            else:
                frequency_dict[item] = 1# create a new key with the item name if the item is found for the first time
    print(frequency_dict)
    
#driver code for above function
add_transaction(['new item'])

{'shrimp': 2, 'almonds': 1, 'avocado': 2, 'vegetables mix': 1, 'green grapes': 1, 'whole wheat flour': 1, 'yams': 1, 'cottage cheese': 1, 'energy drink': 1, 'tomato juice': 1, 'low fat yogurt': 3, 'green tea': 3, 'honey': 3, 'salad': 1, 'mineral water': 6, 'salmon': 3, 'antioxidant juice': 1, 'frozen smoothie': 1, 'spinach': 1, 'olive oil': 1, 'burgers': 2, 'meatballs': 2, 'eggs': 5, 'chutney': 1, 'turkey': 4, 'milk': 2, 'energy bar': 1, 'whole wheat rice': 1, 'whole wheat pasta': 1, 'french fries': 3, 'soup': 1, 'light cream': 1, 'shallot': 1, 'frozen vegetables': 1, 'spaghetti': 3, 'pet food': 1, 'cookies': 2, 'cooking oil': 2, 'champagne': 1, 'chocolate': 1, 'chicken': 2, 'oil': 1, 'fresh tuna': 1, 'tomatoes': 1, 'black tea': 1, 'extra dark chocolate': 1, 'protein bar': 1, 'new item': 1}


# Q4. Walmart is running a discount week for Christmas! The sales manager has decided to keep a 20% discount on the top three selling items. Use the frequency counter that you created in Question 3 to solve this problem:

1. What is the top-selling item?

In [27]:
# Create a list of tuples sorted using the inbuilt function    
listofTuples = sorted(frequency_dict.items() , reverse=True, key=lambda x: x[1])[:1]

listofTuples

[('mineral water', 6)]

In [28]:
#alternate method-by using sorting algorithm
def selection_sort(array):  
    length = len(array)  
      
    for i in range(length-1):  #looping through the length-1 of array
        minIndex = i  
          
        for j in range(i+1, length):  
            if array[minIndex] < array[j]:  
                minIndex = j  
                  
            array[minIndex], array[i] = array[i], array[minIndex] 
          
          
    return array      

array = list(frequency_dict.values())#dictionary values given as list input
  
print("The sorted array is: ", selection_sort(array))  #finding the max value from sorted array

The sorted array is:  [3, 3, 3, 6, 3, 3, 5, 4, 2, 2, 2, 2, 2, 2, 3, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1]


In [29]:
keys = list(frequency_dict.keys()) #creating a variable with list of keys of the dictionary
vals = list(frequency_dict.values()) #creating a variable with list of values of the dictionary
print("Top selling item is",keys[vals.index(6)]) #printing the key for max value found by sorting

Top selling item is mineral water


2. In real life, departmental stores have thousands of items and millions of transactions in their databases. In such situations, it becomes important to use an effective searching algorithm to navigate through the data. You need to sort the frequency counter dictionary by the value count(or the number of items). For example, if there are 3 apples, 5 bananas and 1 watermelon, the sorted list would be banana, apple, watermelon. So you need to sort the entire dictionary (with keys and values) based on the value count. Sort the frequency counter using the sorting algorithms provided in the table below. Additionally, for each of the sorting methods, leverage the time function in Python to see how many milliseconds it takes for an algorithm to sort items.

SELECTION SORT

In [30]:
import time #package to determine the time utilized by the sorting algorithm

t1 = time.time() #time before execution

def selection_sort(array):  
    length = len(array)  
      
    for i in range(length-1):  
        minIndex = i  
          
        for j in range(i+1, length):  
            if array[minIndex] < array[j]:  
                minIndex = j  
                  
            array[minIndex], array[i] = array[i], array[minIndex] 
          
          
    return array      

t2 = time.time() #time after execution
array = list(frequency_dict.values())#dictionary values given as list input
  
print("The sorted array is: ", selection_sort(array))   #printing the sorted array

The sorted array is:  [3, 3, 3, 6, 3, 3, 5, 4, 2, 2, 2, 2, 2, 2, 3, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1]


In [31]:
keys = list(frequency_dict.keys()) #creating a variable with list of keys of the dictionary
vals = list(frequency_dict.values()) #creating a variable with list of values of the dictionary
print("Top 3 items are : ")
print(keys[vals.index(6)])#printing top selling item
print(keys[vals.index(5)])#printing top 2nd selling item
print(keys[vals.index(4)])#printing top 3rd selling item

total_time = (t2-t1)*1000 #formula to calculate time taken by algorithm
print("Time take by selection sort is"'{0:10f}'.format(total_time))

Top 3 items are : 
mineral water
eggs
turkey
Time take by selection sort is  0.092983


INSERTION SORT

In [34]:
t1 = time.time() #before execution
def insertionSort(arr):
  
    # Traverse through 1 to len(arr)
    for i in range(1, len(arr)):
  
        key = arr[i]
  
        # Move elements of arr[0..i-1], that are
        # greater than key, to one position ahead
        # of their current position
        j = i-1
        while j >= 0 and key > arr[j] :
                arr[j + 1] = arr[j]
                j -= 1
        arr[j + 1] = key
        
        
t2 = time.time() #after execution
# Driver code to test above
arr = list(frequency_dict.values()) #dictionary values given as list input
insertionSort(arr) 
print("Sorted array is: ",arr)#printing the sorted array

Sorted array is:  [6, 5, 4, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


In [35]:
keys = list(frequency_dict.keys())#creating a variable with list of keys of the dictionary
vals = list(frequency_dict.values()) #creating a variable with list of values of the dictionary
print("Top 3 items are : ")
print(keys[vals.index(6)])#printing top selling item
print(keys[vals.index(5)])#printing top 2nd selling item
print(keys[vals.index(4)])#printing top 3rd selling item

total_time = (t2-t1)*1000 #formula to calculate time taken by algorithm
print("Time take by insertion sort is"'{0:10f}'.format(total_time))

Top 3 items are : 
mineral water
eggs
turkey
Time take by insertion sort is  0.293732


MERGE SORT

In [36]:
t1 = time.time() #before execution
def mergeSort(arr):
    if len(arr) > 1:

        # Finding the mid of the array
        mid = len(arr)//2

        # Dividing the array elements
        L = arr[:mid]

        # into 2 halves
        R = arr[mid:]

        # Sorting the first half
        mergeSort(L)

        # Sorting the second half
        mergeSort(R)

        i = j = k = 0

        # Copy data to temp arrays L[] and R[]
        while i < len(L) and j < len(R):
            if L[i] > R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1

        # Checking if any element was left
        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1

        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1
            
            
t2 = time.time() #after execution

# Driver Code for above function
arr = list(frequency_dict.values())#dictionary values given as list input
mergeSort(arr)
print("Sorted array is: ", end="\n")
print(arr)#printing the sorted array


Sorted array is: 
[6, 5, 4, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


In [37]:
keys = list(frequency_dict.keys()) #creating the variable for list of keys of the dictionary
vals = list(frequency_dict.values())#creating the variable for list of values of the dictionary
print("Top 3 items are : ")
print(keys[vals.index(6)])#printing top selling item
print(keys[vals.index(5)])#printing top 2nd selling item
print(keys[vals.index(4)])#printing top 3rd selling item

total_time = (t2-t1)*1000 #calculating the time taken by algorithm
print("Time take by merge sort is"'{0:10f}'.format(total_time))

Top 3 items are : 
mineral water
eggs
turkey
Time take by merge sort is  0.382900


3. Rank the sorting algorithms in order of the time they take to sort items.

ANS:
Time take by insertion sort is  0.293732, 
Time take by merge sort is  0.382900 and
Time take by selection sort is  0.092983

Therefore ranking of sorting algorithm is Merge Sort > Insertion Sort > Selection Sort